File: 2020/08_HandheldHalting/instruction.lua
-- ======================================================================
-- Handheld Halting
--   Advent of Code 2020 Day 08 -- Eric Wastl -- https://adventofcode.com
--
-- lua implementation by Dr. Dean Earl Wright III
-- ======================================================================

-- ======================================================================
--                      i n s t r u c t i o n . l u a
-- ======================================================================
-- Instruction for the Advent of Code 2020 Day 08 puzzle

-- ----------------------------------------------------------------------
--                                                                  local
-- ----------------------------------------------------------------------
local Instruction = { part2=false, text='', 
  operation='', argument=0, visited=false}

-- ----------------------------------------------------------------------
--                                                              constants
-- ----------------------------------------------------------------------

-- ======================================================================
--                                                            Instruction
-- ======================================================================


function Instruction:Instruction (o)
  -- Object for Handheld Halting


  -- 1. Set the initial values
  o = o or {}
  o.part2 = o.part2 or false
  o.text = o.text or {}
  o.operation = ''
  o.argument = 0
  o.visited = false

  -- 2. Create the object metatable
  self.__index = self
  setmetatable(o, self)

  -- 3. Process text (if any)
  if o.text ~= nil and #o.text > 0 then
     o:_process_text(o.text)
  end

  return o
end

function Instruction:_process_text(text)
  -- Assign values from text

  -- 0. Precondition axioms
  assert(text ~= nil and #text > 0)
  
  -- 1. Break out the operation and argument
  self.operation = text:match("%a%a%a")
  self.argument = tonumber(text:match("[+-]%d+"))

end

function Instruction:execute(acc)
  -- Execute the instruction, returning the updated accumulator and pc offset
  
  -- 1. Mark instruction as visited
  self.visited = true
  
  -- 2. Handle acc operation
  if self.operation == 'acc' then
    acc = acc + self.argument
    pc = 1
    
  -- 3. Handle jmp operation  
  elseif self.operation == 'jmp' then
    pc = self.argument
    
  -- 4. Handle nop operation  
  elseif self.operation == 'nop' then
    pc = 1
  
  -- 5. Invalid operation
  else
    error("Invalid operation", self.operation)
  end
  
  -- 5. Return the updated accumulator and pc offset
  return acc, pc
end
 
  
-- ----------------------------------------------------------------------
--                                                  module initialization
-- ----------------------------------------------------------------------
return Instruction

-- ======================================================================
-- end                 i n s t r u c t i o n . l u a                  end
-- ======================================================================

File: 2020/08_HandheldHalting/instruction.ts
// ======================================================================
// Handheld Halting
//   Advent of Code 2020 Day 08 -- Eric Wastl -- https://adventofcode.com
//
// TypeScript implementation by Dr. Dean Earl Wright III
// ======================================================================

// ======================================================================
//                           i n s t r u c t i o n . t s
//
// Instruction for the Advent of Code 2020 Day 08 problem
// ======================================================================

// ----------------------------------------------------------------------
//                                                                 import
// ----------------------------------------------------------------------

// ----------------------------------------------------------------------
//                                                              constants
// ----------------------------------------------------------------------
const RE_OP = /[a-z][a-z][a-z]/;
const RE_ARG = /[-+0-9]+/;

// ======================================================================
//                                                            Instruction
// ======================================================================

export class Instruction {
  // Object for Handheld Halting
  text: string;

  part2: boolean;

  operation: string;

  argument: number;

  visited: boolean;

  constructor(text: string, part2 = false) {
    // Create a Instruction object

    // 1. Set the initial values
    this.text = text === undefined ? '' : text;
    this.part2 = part2 === undefined ? false : part2;
    this.operation = '';
    this.argument = 0;
    this.visited = false;

    // 2. Process text (if any)
    if (this.text.length !== 0) {
      let ok = this.text.match(RE_OP);
      if (ok) {
        [this.operation] = ok;
      } else {
        console.log(`'Unable to parse operation from ${this.text}`);
      }
      ok = this.text.match(RE_ARG);
      if (ok) {
        this.argument = parseInt(ok[0], 10);
      } else {
        console.log(`'Unable to parse operation from ${this.text}`);
      }
    }
  }

  execute(acc: number): number[] {
    // Execute the instruction and return the new accumulator and pc delta
    let delta = 1;
    let accumulator = acc;

    // 1. Mark instruction as visited
    this.visited = true;

    // 2. Switch based on the operation
    switch (this.operation) {
      case 'acc':
        accumulator += this.argument;
        break;
      case 'jmp':
        delta = this.argument;
        break;
      case 'nop':
        break;
      default:
        console.log(`Unimplemented operation ${this.operation}`);
    }

    // 3. Return the new accumulator and change to PC
    return [accumulator, delta];
  }

  is_visited(): boolean {
    // Returns true if this instruction has been visited before
    return this.visited;
  }

  reset() {
    // Reset the visited flag
    this.visited = false;
  }
}

// ======================================================================
// end                   i n s t r u c t i o n . t s                  end
// ======================================================================
